FOSUser 2.0

jordisala1991 opened the pull request:
<!-- THE PR TEMPLATE IS NOT AN OPTION. DO NOT DELETE IT, MAKE SURE YOU READ AND EDIT IT! -->

<!--
    Show us you choose the right branch.
    Different branches are used for different things :
    - 3.x is for everything backwards compatible, like patches, features and deprecation notices
    - master is for deprecation removals and other changes that cannot be done without a BC-break
    More details here: https://github.com/sonata-project/SonataUserBundle/blob/3.x/CONTRIBUTING.md#the-base-branch
-->
I am targeting this branch, because this is BC break.

<!--
    Specify which issues will be fixed/closed.
    Remove it if this is not related.
-->

Closes: #794 #779 #718 #702 #672 #666 #322 #379 #563 #553 #444 #410 #409 #745 #627 

Some of them are really really old and maybe even solved right now. Maybe there are more issues that needs to be closed because are talking about 2.x release.

## Changelog

<!-- MANDATORY
    Fill the changelog part inside the code block.
    Follow this schema: http://keepachangelog.com/
-->

<!-- REMOVE EMPTY SECTIONS -->
```markdown
### Added
- Add support for FOSUser 2.0
```

## To do

<!--
    If this is a work in progress, COMPLETE and ADD needed tasks.
    You can add as many tasks as you want.
    If some are not relevant, just REMOVE them.
-->

- [x] Move all code from add_support_for_fos_user2 branch
- [x] Update the documentation
- [x] Add an upgrade note
- [x] Update dev-kit (https://github.com/sonata-project/dev-kit/pull/246)

## Subject

<!-- Describe your Pull Request content here -->
This will bring support for FOSUser 2.0 on master, this will require some changes on 3.x and on dev-kit. There are some code that could be deprecated on 3.x before/after this gets merged.

jordisala1991 commented:
Added a commit with the new travis config, to test with correct versions (will be removed once dev-kit creates its own commit)

jordisala1991 commented:
With this commits, this branch should work like `add_support_for_fos_user2`. Will check it tomorrow.

jontorrado commented:
If PHP7 is needed by default, we should change all `array(` appearances to the square brackets notation @jordisala1991 I'm going to check this as soon as you tell me "PHP stuff done" ;)

jordisala1991 commented:
PHP stuff done @jontorrado 🎉

jordisala1991 commented:
For the `array()` to `[]` I would like to do, but I think we will do it with StyleCI, but since it will make a lot of merge conflicts with 3.x we will do it as late as possible (before releasing 4.x)

jontorrado commented:
@jordisala1991 I will try to check this today, hope to kick all my tasks off!

jontorrado commented:
This is what I found so far:
* The following lines must be added to `SonataUserExtension.php`, line 81:
```
        $tokenStorageReference = new Reference('security.token_storage');
        $authorizationCheckerReference = new Reference('security.authorization_checker');
```
* Documentation (Sonata): new routing imports.
* Documentation (Sonata): Sonata User views removed, so you must move your overriding views to `app/Resources/FOSUserBundle/views/`. Also check namings (some camel case templates changed to camel case)
* Documentation (FOS): Be careful with FOSUser 2.0 removed fields: https://github.com/FriendsOfSymfony/FOSUserBundle/blob/master/Upgrade.md#200-alpha3-to-200-beta1
* Documentation (FOS): form types and handlers changed.

I've been adapting our whole project, I was not able to find any other stuff atm.

jordisala1991 commented:
> The following lines must be added to SonataUserExtension.php, line 81:

Fixed. Removed more lines from SonataUserExtension and moved them to config files.

> Documentation (Sonata): new routing imports.
> Documentation (FOS): Be careful with FOSUser 2.0 removed fields: https://github.com/FriendsOfSymfony/FOSUserBundle/blob/master/Upgrade.md#200-alpha3-to-200-beta1
> Documentation (FOS): form types and handlers changed.

I started with documentation, but there is a lot to do still :)
Handlers were removed on FOSUser, not sure how to handle that on our bundle.

> Documentation (Sonata): Sonata User views removed, so you must move your overriding views to app/Resources/FOSUserBundle/views/. Also check namings (some camel case templates changed to snake case)

This removal will be documented if it is finally done. Not sure if we want to do it or not. (IMO I would remove that code and a lot more)

Thanks a lot for your help :) @jontorrado

jontorrado commented:
@jordisala1991 right now, the following line is not working:
* `$container->setAlias('sonata.user.authentication.form', 'fos_user.profile.form');`

I've also upgraded the whole project to Symfony 3.0.* to see what happens with this pull request. After changing the deprecated stuff, everything is working. I think the most difficult stuff here is the documentation part.

jordisala1991 commented:
@jontorrado Fixed that also.

jontorrado commented:
Upgraded everything to 3.2.*. Fixed all bugs, nothing to do with this fork. Everything is working atm. Not from Sonata User but remember that:
```
twig:
  form:
    resources:
      - 'SonataCoreBundle:Form:datepicker.html.twig'
```
Becomes:
```
twig:
    form_themes:
        - 'SonataCoreBundle:Form:datepicker.html.twig'
```

I'll wait for devkit and further changes. Tell me if I can help with any other stuff.

jontorrado commented:
@jordisala1991 groups are no longer a must in FOSUser2 but, correct me if I'm wrong, SonataUserBundle makes it mandatory in the configuration. We must change this (and also, fix the doc, indeed).

jordisala1991 commented:
Added this commit: https://github.com/sonata-project/SonataUserBundle/pull/869/commits/7f5b6ea36d461bd789afbaa039e135603d05e679

Which removes a lot of code that should be unused after fos controllers were removed.

I will try to make groups optional

jordisala1991 commented:
Added test for the DependencyInjection part. Configuration + Extension.

jordisala1991 commented:
Added a first draft of the UPGRADE note

jontorrado commented:
@jordisala1991 I usually like to add links to every single (outside) documentation. For example, I would add there (inside the upgrade note) the new FOS 2.0 documentation and the upgrade notes: https://github.com/FriendsOfSymfony/FOSUserBundle/blob/master/Upgrade.md. Very important for lazy people to know that items removed and so on ;)

jordisala1991 commented:
Added a link to FOSUser upgrade note.

jordisala1991 commented:
Note that we still need to merge dev-kit PR first. So do not merge this. Only review :)

tomaszlisowiec commented:
Hi guys,
 when this upgrade will be available at packagist ? (even dev)

greg0ire commented:
@tomaszlisowiec Sorry, we don't do estimates for OSS project we work on on our spare time.

jontorrado commented:
@tomaszlisowiec in the meantime, you can add the following to your composer.json:
```
"repositories": [
        {
            "url": "https://github.com/jordisala1991/SonataUserBundle",
            "type": "git"
        }
    ],
```

And in the require section: 
```
        "sonata-project/user-bundle":         "dev-feature/fos_user_2"
```

tomaszlisowiec commented:
Is it the same workaround as this:
composer require sonata-project/user-bundle dev-add_support_for_fos_user2 
?

jordisala1991 commented:
Fixed duplicated flash message when resetting password.
Now Admin controllers do not extend FOSUser ones. They don't need to (not using anything from parent).

> Is it the same workaround as this: composer require sonata-project/user-bundle dev-add_support_for_fos_user2?

It is similar. This PR is from that code. But this one has some improvements and could help us to detect some bugs before merging if you use this branch.

sujayjaju commented:
Is there a reason for PHP7 requirement? FOS supports 5.5

jontorrado commented:
@jordisala1991 I'm checking all the commits daily. No issues found so far. Would you need any help?

jordisala1991 commented:
> Is there a reason for the PHP7 requirement? FOSUser2 supports PHP5.5

Master branch of all sonata bundles will require PHP 7, not sure that we want to do an exception here. PHP 5.5 died some months ago: http://php.net/supported-versions.php

> @jordisala1991 I'm checking all the commits daily. No issues found so far. Would you need any help?

I am still doing some minor changes, but mostly I am waiting for some PR review / approvals from @sonata-project/contributors

jordisala1991 commented:
Fixed some bugs with url generation, there were some using fos_user routes instead of sonata_user_admin ones.

For me this is ready for review. There are some commits not squashed with previous ones on purpose. Because in case we don't want to do that change, I can revert them easily (I will squash them after review).

greg0ire commented on Controller/AdminResettingController.php:
```diff
@@ -11,143 +11,179 @@
+                $user->setLastLogin(new \DateTime());
+            } catch (AccountStatusException $ex) {
+                // We simply do not authenticate users which do not pass the user
+                // checker (not enabled, expired, etc.).
```
Should this exception be logged?

jordisala1991 replied:
Will need to test but, isn't it logged by default on symfony?

jordisala1991 replied:
Do you have any idea on how to trigger such exception?

greg0ire replied:
Why/how would it be logged? It's caught.

jordisala1991 replied:
Right, will add it.

jordisala1991 replied:
Added. It should be enough with a warning right?

greg0ire replied:
Yes, I think it's the right level 👍

jordisala1991 replied:
@greg0ire I am not sure about this one in terms of new dependencies. Should I add monolog to require section? or just check if container has "logger" before calling it out?

greg0ire replied:
The second option seems better

jordisala1991 replied:
Fixed.

jordisala1991 commented on Controller/AdminResettingController.php:
```diff
@@ -11,143 +11,179 @@
-        if (!$user->isPasswordRequestNonExpired($this->container->getParameter('fos_user.resetting.token_ttl'))) {
-            return new RedirectResponse($this->container->get('router')->generate('sonata_user_admin_resetting_request'));
+        if (!$user->isPasswordRequestNonExpired($ttl)) {
+            new RedirectResponse($this->generateUrl('sonata_user_admin_resetting_request'));
```
This should be a return

OskarStark commented on Controller/AdminResettingController.php:
```diff
@@ -11,143 +11,178 @@
+        ));
+
+        // Render the email, use the first line as the subject, and the rest as the body
+        $renderedLines = explode("\n", trim($rendered));
```
would use `PHP_EOL` constant instead of `\n`

Hanmac replied:
i don't know about ``PHP_EOL`` i currently can't test what kind of line sepatator is used for the email template, and can't that be changed by the user?

but what could be done is:
```php
list($subject, $body) = explode("\n", trim($rendered), 2);
```
but i think that does not work if the email template only has one line because with that it does show "Undefined offset"

OskarStark commented on Resources/doc/reference/installation.rst:
```diff
@@ -46,14 +47,7 @@ Next, be sure to enable the bundles in your and ``AppKernel.php`` file:
             new Sonata\BlockBundle\SonataBlockBundle(),
             new Sonata\EasyExtendsBundle\SonataEasyExtendsBundle(),
```
SonataAdminBundle should be enabled here, too

jordisala1991 replied:
Enabled

OskarStark replied:
Thanks

OskarStark commented on Resources/views/Admin/Security/Resetting/reset.html.twig:
```diff
@@ -37,10 +37,10 @@ file that was distributed with this source code.
                 <form action="{{ path('sonata_user_admin_resetting_reset', {'token': token}) }}" method="post" role="form">
                     <div class="form-group">
-                        <input type="password" class="form-control" id="fos_user_resetting_form_new_first"  name="fos_user_resetting_form[new][first]" required="required" placeholder="{{ 'form.new_password'|trans({}, 'FOSUserBundle')|replace({':': ''}) }}"/>
+                        <input type="password" class="form-control" id="fos_user_resetting_form_plainPassword_first" name="fos_user_resetting_form[plainPassword][first]" required="required" placeholder="{{ 'form.new_password'|trans({}, 'FOSUserBundle')|replace({':': ''}) }}"/>
```
no underscore? why do you mix things up here?

jordisala1991 replied:
Just copied this from FOSUserBundle. Not sure if we can make it underscore without changing some FOS code.

jordisala1991 replied:
It does not matter anymore since we are using symfony forms now :D

OskarStark replied:
great

OskarStark commented on Resources/views/Admin/Security/Resetting/reset.html.twig:
```diff
@@ -37,10 +37,10 @@ file that was distributed with this source code.
                     </div>
                     <div class="form-group">
-                        <input type="password" class="form-control" id="fos_user_resetting_form_new_second"  name="fos_user_resetting_form[new][second]" required="required" placeholder="{{ 'form.new_password_confirmation'|trans({}, 'FOSUserBundle')|replace({':': ''}) }}"/>
+                        <input type="password" class="form-control" id="fos_user_resetting_form_plainPassword_second" name="fos_user_resetting_form[plainPassword][second]" required="required" placeholder="{{ 'form.new_password_confirmation'|trans({}, 'FOSUserBundle')|replace({':': ''}) }}"/>
```
same here?

jordisala1991 replied:
symfony forms now

OskarStark replied:
great

Hanmac commented:
about the Resetting Views, is there a reason why we can't use part of FOSUser ones?
or the Form-Types of them, specially 'RegistrationFormType'?

jordisala1991 commented:
> about the Resetting Views, is there a reason why we can't use part of FOSUser ones?
or the Form-Types of them, specially 'RegistrationFormType'?

That part was removed. What do you mean? You can still use FOSUser ones on your own. That code was copy pasted here with a custom integration (to be used on sonata sandbox)

Hanmac commented:
No i mean your change on "Resources/views/Admin/Security/Resetting/reset.html.twig"

1. is that view really needed or can the FOSUser one be used?
1. if that view can't be used, can maybe the Form be used like 'RegistrationFormType'?
(it seems that 'fos_user.resetting.form.factory' does provide it?)

currently "SonataUserBundle:Admin:Security/Resetting/reset.html.twig" does get a form as param, but the view above does not use it.

i don't know, that might be done in a later pull, it doesn't seems that important

jordisala1991 commented:
will take a look at it

jordisala1991 commented:
Fixed, now it uses symfony forms. We need a custom template because we use custom html to display it.

core23 commented:
Just checked the commits, but haven't tried if this really works 😆

OskarStark commented:
What about merging this one and create a 4.0 pre release?

jordisala1991 commented:
Maybe give a final review @greg0ire @core23 @OskarStark and merge? I am okay with a Pre-release, but we will need @Soullivaneuh for that.

greg0ire commented on UPGRADE-4.0.md:
```diff
@@ -12,3 +12,15 @@ See also the [diff code](https://github.com/sonata-project/SonataUserBundle/comp
+
+## Bundle definition
+
+SonataUser no longer have FOSUser as its parent bundle. It requires FOSUser as a dependency always.
```
"have" => "has"

greg0ire commented on UPGRADE-4.0.md:
```diff
@@ -12,3 +12,15 @@ See also the [diff code](https://github.com/sonata-project/SonataUserBundle/comp
+
+## Removals
+
+Removed code used for define custom User login for the fronted application. Use FOSUser directly instead.
```
- "for define" => "to define"
- "fronted" => "frontend"

jordisala1991 replied:
Fixed both

OskarStark commented:
LGTM, I would merge it this way

jordisala1991 commented:
![87xihbthj1dka](https://cloud.githubusercontent.com/assets/1137485/25793985/47d3a0da-33cf-11e7-91f1-3a55e2f54304.gif)

OskarStark commented:
Thanks @jordisala1991 🎉 🎉 🎉

ceesvanegmond commented:
@OskarStark tag tag tag 👍

OskarStark commented:
I requested it internally

jontorrado commented:
Awesome work @jordisala1991

greg0ire commented:
Congrats on finishing this!

florimond83 commented:
Thanks a lot @jordisala1991 for the work! Just wanted to signal an error on my attempt, under php 7.0.18 and tested as well with 7.1.4, I get the same issue trying to install, don't know if it is due to some emulation for Mac OS Sierra (I've followed to upgrade PHP https://coolestguidesontheplanet.com/upgrade-php-on-osx/) : 

On executing : 'composer require sonata-project/user-bundle:4.x-dev'
I got:
> Your requirements could not be resolved to an installable set of packages.
> 
>   Problem 1
>     - Installation request for sonata-project/user-bundle 4.x-dev -> satisfiable by sonata-project/user-bundle[4.x-dev].
>     - sonata-project/user-bundle 4.x-dev requires php ^7.0 -> your PHP version (7.0.18) overridden by "config.platform.php" version (5.5.9) does not satisfy that requirement.

Thus I used "--ignore-platform-reqs" and it looks like ok...still wanted to share.

OskarStark commented:
did you set this in your composer.json @florimond83 :
`config.platform.php" version (5.5.9)`

OskarStark commented:
I closed the referenced tickets manually

florimond83 commented:
Hi @OskarStark, no haven't touched  composer.json, I'm using a fresh new Symfony 3.2.8 install with fos user 2.0.
I did use --ignore-platform-reqs and it worked.

Hanmac commented:
@OskarStark https://github.com/sonata-project/SonataUserBundle/issues/877 should probably be closed too right?

vyshkant commented:
As I understand, thees changes are not released yet?

I'm trying to `require` SonataUserBundle and it fails:
```
$ composer require sonata-project/user-bundle
Using version ^3.2 for sonata-project/user-bundle
./composer.json has been updated
Loading composer repositories with package information
Updating dependencies (including require-dev)
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - Conclusion: remove symfony/symfony v3.2.8
    - Conclusion: don't install symfony/symfony v3.2.8
    - sonata-project/user-bundle 3.2.0 requires symfony/form ^2.3 -> satisfiable by symfony/form[v2.3.0, v2.3.1, v2.3.10, v2.3.11, v2.3.12, v2.3.13, v2.3.14, v2.3.15, v2.3.16, v2.3.17, v2.3.18, v2.3.19, v2.3.2, v2.3.20, v2.3.21, v2.3.22, v2.3.23, v2.3.24, v2.3.25, v2.3.26, v2.3.27, v2.3.28, v2.3.29, v2.3.3, v2.3.30, v2.3.31, v2.3.32, v2.3.33, v2.3.34, v2.3.35, v2.3.36, v2.3.37, v2.3.38, v2.3.39, v2.3.4, v2.3.40, v2.3.41, v2.3.42, v2.3.5, v2.3.6, v2.3.7, v2.3.8, v2.3.9, v2.4.0, v2.4.1, v2.4.10, v2.4.2, v2.4.3, v2.4.4, v2.4.5, v2.4.6, v2.4.7, v2.4.8, v2.4.9, v2.5.0, v2.5.1, v2.5.10, v2.5.11, v2.5.12, v2.5.2, v2.5.3, v2.5.4, v2.5.5, v2.5.6, v2.5.7, v2.5.8, v2.5.9, v2.6.0, v2.6.1, v2.6.10, v2.6.11, v2.6.12, v2.6.13, v2.6.2, v2.6.3, v2.6.4, v2.6.5, v2.6.6, v2.6.7, v2.6.8, v2.6.9, v2.7.0, v2.7.1, v2.7.10, v2.7.11, v2.7.12, v2.7.13, v2.7.14, v2.7.15, v2.7.16, v2.7.17, v2.7.18, v2.7.19, v2.7.2, v2.7.20, v2.7.21, v2.7.22, v2.7.23, v2.7.24, v2.7.25, v2.7.26, v2.7.27, v2.7.3, v2.7.4, v2.7.5, v2.7.6, v2.7.7, v2.7.8, v2.7.9, v2.8.0, v2.8.1, v2.8.10, v2.8.11, v2.8.12, v2.8.13, v2.8.14, v2.8.15, v2.8.16, v2.8.17, v2.8.18, v2.8.19, v2.8.2, v2.8.20, v2.8.3, v2.8.4, v2.8.5, v2.8.6, v2.8.7, v2.8.8, v2.8.9].
    - sonata-project/user-bundle 3.2.1 requires symfony/form ^2.3 -> satisfiable by symfony/form[v2.3.0, v2.3.1, v2.3.10, v2.3.11, v2.3.12, v2.3.13, v2.3.14, v2.3.15, v2.3.16, v2.3.17, v2.3.18, v2.3.19, v2.3.2, v2.3.20, v2.3.21, v2.3.22, v2.3.23, v2.3.24, v2.3.25, v2.3.26, v2.3.27, v2.3.28, v2.3.29, v2.3.3, v2.3.30, v2.3.31, v2.3.32, v2.3.33, v2.3.34, v2.3.35, v2.3.36, v2.3.37, v2.3.38, v2.3.39, v2.3.4, v2.3.40, v2.3.41, v2.3.42, v2.3.5, v2.3.6, v2.3.7, v2.3.8, v2.3.9, v2.4.0, v2.4.1, v2.4.10, v2.4.2, v2.4.3, v2.4.4, v2.4.5, v2.4.6, v2.4.7, v2.4.8, v2.4.9, v2.5.0, v2.5.1, v2.5.10, v2.5.11, v2.5.12, v2.5.2, v2.5.3, v2.5.4, v2.5.5, v2.5.6, v2.5.7, v2.5.8, v2.5.9, v2.6.0, v2.6.1, v2.6.10, v2.6.11, v2.6.12, v2.6.13, v2.6.2, v2.6.3, v2.6.4, v2.6.5, v2.6.6, v2.6.7, v2.6.8, v2.6.9, v2.7.0, v2.7.1, v2.7.10, v2.7.11, v2.7.12, v2.7.13, v2.7.14, v2.7.15, v2.7.16, v2.7.17, v2.7.18, v2.7.19, v2.7.2, v2.7.20, v2.7.21, v2.7.22, v2.7.23, v2.7.24, v2.7.25, v2.7.26, v2.7.27, v2.7.3, v2.7.4, v2.7.5, v2.7.6, v2.7.7, v2.7.8, v2.7.9, v2.8.0, v2.8.1, v2.8.10, v2.8.11, v2.8.12, v2.8.13, v2.8.14, v2.8.15, v2.8.16, v2.8.17, v2.8.18, v2.8.19, v2.8.2, v2.8.20, v2.8.3, v2.8.4, v2.8.5, v2.8.6, v2.8.7, v2.8.8, v2.8.9].
    - sonata-project/user-bundle 3.2.2 requires symfony/form ^2.3 -> satisfiable by symfony/form[v2.3.0, v2.3.1, v2.3.10, v2.3.11, v2.3.12, v2.3.13, v2.3.14, v2.3.15, v2.3.16, v2.3.17, v2.3.18, v2.3.19, v2.3.2, v2.3.20, v2.3.21, v2.3.22, v2.3.23, v2.3.24, v2.3.25, v2.3.26, v2.3.27, v2.3.28, v2.3.29, v2.3.3, v2.3.30, v2.3.31, v2.3.32, v2.3.33, v2.3.34, v2.3.35, v2.3.36, v2.3.37, v2.3.38, v2.3.39, v2.3.4, v2.3.40, v2.3.41, v2.3.42, v2.3.5, v2.3.6, v2.3.7, v2.3.8, v2.3.9, v2.4.0, v2.4.1, v2.4.10, v2.4.2, v2.4.3, v2.4.4, v2.4.5, v2.4.6, v2.4.7, v2.4.8, v2.4.9, v2.5.0, v2.5.1, v2.5.10, v2.5.11, v2.5.12, v2.5.2, v2.5.3, v2.5.4, v2.5.5, v2.5.6, v2.5.7, v2.5.8, v2.5.9, v2.6.0, v2.6.1, v2.6.10, v2.6.11, v2.6.12, v2.6.13, v2.6.2, v2.6.3, v2.6.4, v2.6.5, v2.6.6, v2.6.7, v2.6.8, v2.6.9, v2.7.0, v2.7.1, v2.7.10, v2.7.11, v2.7.12, v2.7.13, v2.7.14, v2.7.15, v2.7.16, v2.7.17, v2.7.18, v2.7.19, v2.7.2, v2.7.20, v2.7.21, v2.7.22, v2.7.23, v2.7.24, v2.7.25, v2.7.26, v2.7.27, v2.7.3, v2.7.4, v2.7.5, v2.7.6, v2.7.7, v2.7.8, v2.7.9, v2.8.0, v2.8.1, v2.8.10, v2.8.11, v2.8.12, v2.8.13, v2.8.14, v2.8.15, v2.8.16, v2.8.17, v2.8.18, v2.8.19, v2.8.2, v2.8.20, v2.8.3, v2.8.4, v2.8.5, v2.8.6, v2.8.7, v2.8.8, v2.8.9].
    - sonata-project/user-bundle 3.2.3 requires symfony/form ^2.3 -> satisfiable by symfony/form[v2.3.0, v2.3.1, v2.3.10, v2.3.11, v2.3.12, v2.3.13, v2.3.14, v2.3.15, v2.3.16, v2.3.17, v2.3.18, v2.3.19, v2.3.2, v2.3.20, v2.3.21, v2.3.22, v2.3.23, v2.3.24, v2.3.25, v2.3.26, v2.3.27, v2.3.28, v2.3.29, v2.3.3, v2.3.30, v2.3.31, v2.3.32, v2.3.33, v2.3.34, v2.3.35, v2.3.36, v2.3.37, v2.3.38, v2.3.39, v2.3.4, v2.3.40, v2.3.41, v2.3.42, v2.3.5, v2.3.6, v2.3.7, v2.3.8, v2.3.9, v2.4.0, v2.4.1, v2.4.10, v2.4.2, v2.4.3, v2.4.4, v2.4.5, v2.4.6, v2.4.7, v2.4.8, v2.4.9, v2.5.0, v2.5.1, v2.5.10, v2.5.11, v2.5.12, v2.5.2, v2.5.3, v2.5.4, v2.5.5, v2.5.6, v2.5.7, v2.5.8, v2.5.9, v2.6.0, v2.6.1, v2.6.10, v2.6.11, v2.6.12, v2.6.13, v2.6.2, v2.6.3, v2.6.4, v2.6.5, v2.6.6, v2.6.7, v2.6.8, v2.6.9, v2.7.0, v2.7.1, v2.7.10, v2.7.11, v2.7.12, v2.7.13, v2.7.14, v2.7.15, v2.7.16, v2.7.17, v2.7.18, v2.7.19, v2.7.2, v2.7.20, v2.7.21, v2.7.22, v2.7.23, v2.7.24, v2.7.25, v2.7.26, v2.7.27, v2.7.3, v2.7.4, v2.7.5, v2.7.6, v2.7.7, v2.7.8, v2.7.9, v2.8.0, v2.8.1, v2.8.10, v2.8.11, v2.8.12, v2.8.13, v2.8.14, v2.8.15, v2.8.16, v2.8.17, v2.8.18, v2.8.19, v2.8.2, v2.8.20, v2.8.3, v2.8.4, v2.8.5, v2.8.6, v2.8.7, v2.8.8, v2.8.9].
    - don't install symfony/form v2.3.0|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.3.1|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.3.10|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.3.11|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.3.12|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.3.13|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.3.14|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.3.15|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.3.16|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.3.17|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.3.18|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.3.19|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.3.2|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.3.20|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.3.21|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.3.22|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.3.23|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.3.24|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.3.25|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.3.26|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.3.27|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.3.28|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.3.29|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.3.3|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.3.30|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.3.31|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.3.32|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.3.33|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.3.34|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.3.35|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.3.36|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.3.37|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.3.38|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.3.39|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.3.4|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.3.40|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.3.41|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.3.42|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.3.5|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.3.6|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.3.7|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.3.8|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.3.9|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.4.0|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.4.1|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.4.10|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.4.2|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.4.3|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.4.4|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.4.5|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.4.6|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.4.7|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.4.8|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.4.9|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.5.0|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.5.1|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.5.10|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.5.11|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.5.12|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.5.2|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.5.3|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.5.4|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.5.5|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.5.6|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.5.7|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.5.8|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.5.9|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.6.0|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.6.1|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.6.10|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.6.11|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.6.12|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.6.13|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.6.2|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.6.3|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.6.4|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.6.5|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.6.6|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.6.7|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.6.8|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.6.9|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.7.0|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.7.1|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.7.10|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.7.11|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.7.12|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.7.13|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.7.14|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.7.15|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.7.16|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.7.17|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.7.18|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.7.19|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.7.2|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.7.20|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.7.21|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.7.22|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.7.23|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.7.24|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.7.25|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.7.26|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.7.27|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.7.3|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.7.4|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.7.5|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.7.6|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.7.7|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.7.8|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.7.9|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.8.0|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.8.1|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.8.10|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.8.11|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.8.12|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.8.13|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.8.14|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.8.15|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.8.16|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.8.17|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.8.18|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.8.19|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.8.2|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.8.20|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.8.3|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.8.4|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.8.5|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.8.6|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.8.7|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.8.8|don't install symfony/symfony v3.2.8
    - don't install symfony/form v2.8.9|don't install symfony/symfony v3.2.8
    - Installation request for symfony/symfony (locked at v3.2.8, required as 3.2.*) -> satisfiable by symfony/symfony[v3.2.8].
    - Installation request for sonata-project/user-bundle ^3.2 -> satisfiable by sonata-project/user-bundle[3.2.0, 3.2.1, 3.2.2, 3.2.3].


Installation failed, reverting ./composer.json to its original content.
```

My environment:

#### Sonata packages

```bash
$ composer show sonata-project/*
sonata-project/admin-bundle              3.17.0 The missing Symfony Admin Generator
sonata-project/block-bundle              3.3.2  Symfony SonataBlockBundle
sonata-project/cache                     1.0.7  Cache library
sonata-project/core-bundle               3.3.0  Symfony SonataCoreBundle
sonata-project/datagrid-bundle           2.2.1  Symfony SonataDatagridBundle
sonata-project/doctrine-extensions       1.0.2  Doctrine2 behavioral extensions
sonata-project/doctrine-orm-admin-bundle 3.1.4  Symfony Sonata / Integrate Doctrine ORM into the SonataAdminBundle
sonata-project/easy-extends-bundle       2.2.0  Symfony SonataEasyExtendsBundle
sonata-project/exporter                  1.7.1  Lightweight Exporter library
sonata-project/media-bundle              3.5.1  Symfony SonataMediaBundle
sonata-project/notification-bundle       3.1.0  Symfony SonataNotificationBundle
```

#### Symfony packages

```bash
$ composer show symfony/*
symfony/assetic-bundle     v2.8.1 Integrates Assetic into Symfony2
symfony/monolog-bundle     v3.1.0 Symfony MonologBundle
symfony/phpunit-bridge     v3.2.8 Symfony PHPUnit Bridge
symfony/polyfill-apcu      v1.3.0 Symfony polyfill backporting apcu_* functions to lower PHP versions
symfony/polyfill-intl-icu  v1.3.0 Symfony polyfill for intl's ICU-related data and classes
symfony/polyfill-mbstring  v1.3.0 Symfony polyfill for the Mbstring extension
symfony/polyfill-php56     v1.3.0 Symfony polyfill backporting some PHP 5.6+ features to lower PHP versions
symfony/polyfill-php70     v1.3.0 Symfony polyfill backporting some PHP 7.0+ features to lower PHP versions
symfony/polyfill-util      v1.3.0 Symfony utilities for portability of PHP codes
symfony/security-acl       v3.0.0 Symfony Security Component - ACL (Access Control List)
symfony/swiftmailer-bundle v2.5.4 Symfony SwiftmailerBundle
symfony/symfony            v3.2.8 The Symfony PHP framework
```

#### PHP version

```bash
$ php -v
PHP 7.1.5-1+deb.sury.org~xenial+1 (cli) (built: May 11 2017 14:07:52) ( NTS )
Copyright (c) 1997-2017 The PHP Group
Zend Engine v3.1.0, Copyright (c) 1998-2017 Zend Technologies
    with Zend OPcache v7.1.5-1+deb.sury.org~xenial+1, Copyright (c) 1999-2017, by Zend Technologies
    with Xdebug v2.5.1, Copyright (c) 2002-2017, by Derick Rethans
```

rechengehirn commented:
>   Problem 1
>     - sonata-project/user-bundle 4.x-dev conflicts with sonata-project/intl-bundle[2.3.0].
>     - sonata-project/user-bundle 4.x-dev conflicts with sonata-project/intl-bundle[2.3.0].
>     - sonata-project/user-bundle 4.x-dev conflicts with sonata-project/intl-bundle[2.3.0].
>     - Installation request for sonata-project/user-bundle 4.x-dev -> satisfiable by sonata-project/user-bundle[4.x-dev].
>     - Installation request for sonata-project/intl-bundle ^2.3 -> satisfiable by sonata-project/intl-bundle[2.3.0].

Can someone help with that?

jordisala1991 commented:
Here is the problem. https://github.com/sonata-project/SonataIntlBundle/blob/2.x/composer.json#L37.

First we need to see why is that conflict there. Can you take a look at intl-bundle code to see if there is any relation with SonataUser?

ceesvanegmond commented:
Any roadmap available for version 4.0?

greg0ire commented:
@ceesvanegmond yes : https://github.com/sonata-project/dev-kit/issues/216
